File: ui/src/api/kds.ts
import { loggedInvoke } from '@/utils/logged-invoke';

/** Status of a Kitchen Display System order. */
export type KdsStatus = 'pending' | 'preparing' | 'ready' | 'served' | 'cancelled';

/** A modifier choice attached to a line item. */
export interface KdsModifier {
  name: string;
  choice: string;
  price_minor: number;
}

/** A single line item on a KDS order ticket. */
export interface KdsLineItem {
  id: string;
  kds_order_id: string;
  sku: string;
  display_name: string;
  qty: number;
  /** Course assignment: "appetizer", "main", "dessert", "beverage", or null. */
  course: string | null;
  /** Modifier choices. */
  modifiers: KdsModifier[];
  line_position: number;
  item_status: string;
  started_at: string | null;
  ready_at: string | null;
  served_at: string | null;
  created_at: string;
}

/** A Kitchen Display System order. */
export interface KdsOrder {
  id: string;
  sale_id: string;
  /** The store this order belongs to (ADR #8). Used for defense-in-depth filtering. */
  store_id: string | null;
  status: KdsStatus;
  items_summary: string;
  item_count: number;
  display_number: number | null;
  received_at: string;
  started_at: string | null;
  ready_at: string | null;
  served_at: string | null;
  prep_time_seconds: number;
  /** Kitchen zone this order is assigned to (e.g., "front", "back"). */
  kitchen_zone: string | null;
  notes: string;
  /** Table number assigned to this order (e.g., "T5"), or null for takeaway. */
  table_number: string | null;
  /** Priority/rush flag: when true the ticket visually escalates above normal SLA. */
  priority: boolean;
}

/** List KDS orders, optionally filtered by status. */
export const listKdsOrders = (userId: string, status?: KdsStatus): Promise<KdsOrder[]> =>
  loggedInvoke<KdsOrder[]>('list_kds_orders', { userId, status: status ?? null });

/** List KDS orders (scoped — ADR #7). */
export const listKdsOrdersScoped = (sessionToken: string, status?: KdsStatus): Promise<KdsOrder[]> =>
  loggedInvoke<KdsOrder[]>('list_kds_orders_scoped', { sessionToken, status: status ?? null });

/** Get the KDS queue for the current user, optionally filtered by kitchen zone. */
export const getKdsQueue = (userId: string, kdsZone?: string): Promise<KdsOrder[]> =>
  loggedInvoke<KdsOrder[]>('get_kds_queue', { userId, kdsZone: kdsZone ?? null });

/** Get the KDS queue (scoped — ADR #7), optionally filtered by kitchen zone. */
export const getKdsQueueScoped = (sessionToken: string, kdsZone?: string): Promise<KdsOrder[]> =>
  loggedInvoke<KdsOrder[]>('get_kds_queue_scoped', { sessionToken, kdsZone: kdsZone ?? null });

/** Update a KDS order's status (e.g. preparing, ready, served). */
export const updateKdsStatus = (userId: string, id: string, status: KdsStatus): Promise<KdsOrder> =>
  loggedInvoke<KdsOrder>('update_kds_status', { userId, id, status });

/** Update a KDS order's status (scoped — ADR #7). */
export const updateKdsStatusScoped = (sessionToken: string, id: string, status: KdsStatus): Promise<KdsOrder> =>
  loggedInvoke<KdsOrder>('update_kds_status_scoped', { sessionToken, id, status });

/** Create KDS orders from a completed sale. Returns one order per kitchen zone. */
export const createKdsOrderFromSale = (userId: string, saleId: string): Promise<KdsOrder[]> =>
  loggedInvoke<KdsOrder[]>('create_kds_order_from_sale', { userId, saleId });

/** Create KDS orders from a sale (scoped — ADR #7). */
export const createKdsOrderFromSaleScoped = (sessionToken: string, saleId: string): Promise<KdsOrder[]> =>
  loggedInvoke<KdsOrder[]>('create_kds_order_from_sale_scoped', { sessionToken, saleId });

/** Get a single KDS order by its identifier. */
export const getKdsOrder = (userId: string, id: string): Promise<KdsOrder | null> =>
  loggedInvoke<KdsOrder | null>('get_kds_order', { userId, id });

/** Get a KDS order by id (scoped — ADR #7). */
export const getKdsOrderScoped = (sessionToken: string, id: string): Promise<KdsOrder | null> =>
  loggedInvoke<KdsOrder | null>('get_kds_order_scoped', { sessionToken, id });

/** Input for creating a KDS line item (mirrors Rust CreateKdsLineItemInput). */
export interface CreateKdsLineItemInput {
  sku: string;
  display_name: string;
  qty: number;
  course: string | null;
  modifiers: KdsModifier[];
}

/** Input for updating items on an existing KDS order. */
export interface UpdateKdsOrderItemsInput {
  id: string;
  items_summary: string;
  item_count: number;
  /** Structured line items to replace kds_line_items. When provided, summary/count are re-derived. */
  line_items?: CreateKdsLineItemInput[] | null;
}

/** Update the items (summary + count) on an existing KDS order. */
export const updateKdsOrderItems = (userId: string, args: UpdateKdsOrderItemsInput): Promise<KdsOrder> =>
  loggedInvoke<KdsOrder>('update_kds_order_items', { userId, args });

/** Update KDS order items (scoped — ADR #7). */
export const updateKdsOrderItemsScoped = (sessionToken: string, args: UpdateKdsOrderItemsInput): Promise<KdsOrder> =>
  loggedInvoke<KdsOrder>('update_kds_order_items_scoped', { sessionToken, args });

/** Print a kitchen chit for a KDS order (scoped — ADR #7). */
export const printKdsChitScoped = (sessionToken: string, orderId: string): Promise<boolean> =>
  loggedInvoke<boolean>('print_kds_chit_scoped', { sessionToken, orderId });

/** Get all line items for a KDS order (scoped — ADR #7). */
export const getKdsOrderLinesScoped = (sessionToken: string, orderId: string): Promise<KdsLineItem[]> =>
  loggedInvoke<KdsLineItem[]>('get_kds_order_lines_scoped', { sessionToken, orderId });

/** Update the status of a single KDS line item (scoped — ADR #7). */
export const updateKdsLineItemStatusScoped = (
  sessionToken: string,
  itemId: string,
  status: KdsStatus,
): Promise<KdsLineItem> =>
  loggedInvoke<KdsLineItem>('update_kds_line_item_status_scoped', { sessionToken, itemId, status });

File: crates/oz-core/src/kds.rs
//! Kitchen Display System (KDS) domain types.
//!
//! Types for order tickets that route completed sales to the kitchen
//! display system with status tracking and timestamps.

use serde::{Deserialize, Serialize};

/// Status of a KDS order in the kitchen workflow.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum KdsStatus {
    /// Order received, not yet being worked on.
    Pending,
    /// Kitchen is actively preparing the order.
    Preparing,
    /// Order is ready to be served.
    Ready,
    /// Order has been served to the customer.
    Served,
    /// Order was cancelled.
    Cancelled,
}

impl KdsStatus {
    /// Serialize to the database string representation.
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Pending => "pending",
            Self::Preparing => "preparing",
            Self::Ready => "ready",
            Self::Served => "served",
            Self::Cancelled => "cancelled",
        }
    }

    /// Parse from a database string representation.
    #[allow(clippy::should_implement_trait)]
    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "pending" => Some(Self::Pending),
            "preparing" => Some(Self::Preparing),
            "ready" => Some(Self::Ready),
            "served" => Some(Self::Served),
            "cancelled" => Some(Self::Cancelled),
            _ => None,
        }
    }
}

/// A KDS order ticket displayed in the kitchen.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KdsOrder {
    /// Primary key (UUID v4).
    pub id: String,
    /// FK to the originating sale.
    pub sale_id: String,
    /// The store where the order belongs (ADR #8).
    ///
    /// Populated from the sale's store context. Used by KDS tablets
    /// to filter orders for defense-in-depth in multi-store deployments.
    pub store_id: Option<String>,
    /// Topology-selected KDS workspace instance for this ticket.
    ///
    /// `None` is retained for tickets created before runtime route
    /// compilation or by legacy unscoped callers.
    #[serde(default)]
    pub target_instance_id: Option<String>,
    /// Current kitchen status ("pending", "preparing", "ready", "served", "cancelled").
    pub status: String,
    /// Comma-separated item names for display.
    pub items_summary: String,
    /// Total number of items in the order.
    pub item_count: i64,
    /// Human-readable display number (auto-increment per day).
    pub display_number: Option<i64>,
    /// ISO-8601 timestamp of when the order was received.
    pub received_at: String,
    /// ISO-8601 timestamp of when preparation started.
    pub started_at: Option<String>,
    /// ISO-8601 timestamp of when preparation finished.
    pub ready_at: Option<String>,
    /// ISO-8601 timestamp of when the order was served.
    pub served_at: Option<String>,
    /// Estimated preparation time in seconds.
    pub prep_time_seconds: i64,
    /// Kitchen zone this order belongs to (e.g., "front", "back").
    ///
    /// Populated from the product's `kitchen_zone` at sale completion time.
    /// Used by KDS devices to filter their queue to only their assigned zone.
    pub kitchen_zone: Option<String>,
    /// Special notes from the POS (e.g., "no onions").
    pub notes: String,
    /// Table number assigned to this order (e.g., "T5").
    ///
    /// Populated from the `tables` table at order-creation time via
    /// the sale's `active_sale_id` link. `None` for takeaway orders.
    pub table_number: Option<String>,
    /// Priority/rush flag: when true the ticket visually escalates above normal SLA.
    /// Set by FOH to signal an urgent order (e.g., VIP, long wait, special request).
    pub priority: bool,
}

/// A modifier choice attached to a line item.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct KdsModifier {
    /// Modifier group name (e.g., "Temperature", "Add-ons").
    pub name: String,
    /// Selected option (e.g., "Medium Rare", "Extra Cheese").
    pub choice: String,
    /// Price impact in minor units (0 when included).
    #[serde(default)]
    pub price_minor: i64,
}

/// A single line item on a KDS order ticket.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KdsLineItem {
    /// Primary key (UUIDv7).
    pub id: String,
    /// FK to the parent KDS order.
    pub kds_order_id: String,
    /// Product SKU.
    pub sku: String,
    /// Product display name (resolved at creation time).
    pub display_name: String,
    /// Quantity (≥ 1).
    pub qty: i64,
    /// Course assignment ("appetizer", "main", "dessert", "beverage", or NULL).
    pub course: Option<String>,
    /// Modifier choices (empty vec when no modifiers).
    #[serde(default)]
    pub modifiers: Vec<KdsModifier>,
    /// Display order within the ticket.
    pub line_position: i64,
    /// Per-item status.
    #[serde(default)]
    pub item_status: String,
    /// ISO-8601 timestamp of when preparation started.
    pub started_at: Option<String>,
    /// ISO-8601 timestamp of when preparation finished.
    pub ready_at: Option<String>,
    /// ISO-8601 timestamp of when the item was served.
    pub served_at: Option<String>,
    /// ISO-8601 creation timestamp.
    pub created_at: String,
}

/// Input for creating a KDS order from a completed sale.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateKdsOrderInput {
    /// FK to the originating sale.
    pub sale_id: String,
    /// The store where this order belongs (ADR #8).
    pub store_id: Option<String>,
    /// Derived flat summary (e.g. "Steak x2, Salad") — populated from items.
    pub items_summary: String,
    /// Total item count — derived from items.
    pub item_count: i64,
    /// Kitchen zone to assign (e.g., "front", "back").
    pub kitchen_zone: Option<String>,
    /// Special notes.
    pub notes: String,
    /// Table number assigned to this order (e.g., "T5").
    pub table_number: Option<String>,
    /// Priority/rush flag: when true the ticket visually escalates above normal SLA.
    pub priority: bool,
}

/// Input for creating a KDS line item.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateKdsLineItemInput {
    /// Product SKU.
    pub sku: String,
    /// Product display name.
    pub display_name: String,
    /// Quantity (≥ 1).
    pub qty: i64,
    /// Course assignment.
    pub course: Option<String>,
    /// Modifier choices.
    pub modifiers: Vec<KdsModifier>,
}

/// Input for updating the items on an existing KDS order.
///
/// Used when FOH adds items to an order mid-preparation, or when
/// kitchen staff need to correct the items shown on a ticket.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateKdsOrderItemsInput {
    /// KDS order ID to update.
    pub id: String,
    /// Updated comma-separated item display names.
    pub items_summary: String,
    /// Updated total item count.
    pub item_count: i64,
    /// Structured line items to replace the existing kds_line_items.
    ///
    /// When `Some`, the existing line items are deleted and replaced
    /// with these. The `items_summary` and `item_count` fields are
    /// re-derived from this data (the string/count inputs are ignored).
    /// When `None`, only the summary/count are updated (legacy behaviour).
    #[serde(default)]
    pub line_items: Option<Vec<CreateKdsLineItemInput>>,
}

/// A registered KDS display device bound to one Restaurant POS.
///
/// Each device is enrolled via QR-code pairing and tracked through
/// the `kds_devices` table. The `station_ids` field determines which
/// topology stations this device is responsible for — an empty vec
/// means the device receives all orders (broadcast mode).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct KdsDevice {
    /// Unique device identifier (UUID v7).
    pub id: String,
    /// Human-readable display name (e.g. "Expo Screen").
    pub name: String,
    /// FK to the parent Restaurant POS terminal.
    pub restaurant_pos_id: String,
    /// Topology station IDs this device is responsible for.
    /// Empty vec = receives all orders (broadcast mode).
    pub station_ids: Vec<String>,
    /// Whether this device is currently active/enrolled.
    pub is_active: bool,
    /// ISO-8601 timestamp of last communication, `None` if never connected.
    pub last_seen_at: Option<String>,
    /// Current connection status.
    pub connection_status: KdsConnectionStatus,
    /// ISO-8601 creation timestamp.
    pub created_at: String,
    /// ISO-8601 last-update timestamp.
    pub updated_at: String,
}

/// Connection status of a KDS device.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum KdsConnectionStatus {
    /// Device is actively connected and receiving events.
    Connected,
    /// Device is not currently connected.
    Disconnected,
    /// Device was connected but has not communicated recently.
    Stale,
}

impl KdsConnectionStatus {
    /// Serialize to the database string representation.
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Connected => "connected",
            Self::Disconnected => "disconnected",
            Self::Stale => "stale",
        }
    }

    /// Parse from a database string representation.
    pub fn parse_db(s: &str) -> Option<Self> {
        match s {
            "connected" => Some(Self::Connected),
            "disconnected" => Some(Self::Disconnected),
            "stale" => Some(Self::Stale),
            _ => None,
        }
    }
}

/// Input for registering a new KDS device.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegisterKdsDeviceInput {
    /// Display name for the device.
    pub name: String,
    /// The Restaurant POS terminal ID this device is bound to.
    pub restaurant_pos_id: String,
    /// Topology station IDs this device is responsible for.
    pub station_ids: Vec<String>,
    /// SHA-256 hash of the enrollment token.
    pub pairing_token_hash: String,
    /// ISO-8601 expiry timestamp for the enrollment token.
    pub pairing_expires_at: String,
}

/// Resolve which KDS devices should receive an order based on its line items.
///
/// 1. For each line item, look up its product's topology station assignment.
/// 2. Match station → KDS devices via `kds_devices.station_ids`.
/// 3. If a device has an empty `station_ids` (broadcast mode), it receives all orders.
/// 4. If no device claims a station, the order broadcasts to all devices (safe fallback).
/// 5. Deduplicate — a device never receives the same order twice.
///
/// `station_for_sku` is a callback that maps a SKU to its topology station ID.
/// Return `None` if the SKU has no station assignment.
pub fn resolve_kds_targets<F>(
    line_items: &[KdsLineItem],
    devices: &[KdsDevice],
    station_for_sku: F,
) -> Vec<String>
where
    F: Fn(&str) -> Option<String>,
{
    use std::collections::HashSet;

    let mut targeted_devices: HashSet<String> = HashSet::new();
    let mut untargeted_stations: HashSet<String> = HashSet::new();

    // Phase 1: Station-based targeting
    for item in line_items {
        if let Some(station) = station_for_sku(&item.sku) {
            let mut any_device_claimed = false;
            for device in devices {
                if device.is_active && device.station_ids.contains(&station) {
                    targeted_devices.insert(device.id.clone());
                    any_device_claimed = true;
                }
            }
            if !any_device_claimed {
                untargeted_stations.insert(station);
            }
        }
    }

    // Phase 2: Broadcast fallback — empty station_ids means "show everything"
    for device in devices {
        if device.is_active && device.station_ids.is_empty() {
            targeted_devices.insert(device.id.clone());
        }
    }

    // Phase 3: If any station has no claiming device, broadcast to all
    if !untargeted_stations.is_empty() {
        for device in devices {
            if device.is_active {
                targeted_devices.insert(device.id.clone());
            }
        }
    }

    targeted_devices.into_iter().collect()
}

#[cfg(test)]
#[path = "kds_tests.rs"]
mod tests;
